File: bindings/kotlin/src/main/kotlin/com/elabify/core/Did.kt
// did:elabify parsing and formatting. Byte-equivalent with @elabify/core's
// src/did.ts. Grammar (exactly five colon-separated components):
//
//   did:elabify:<network>:<entityType>:<identifier>
//
// Cross-binding equivalence: test-vectors/did.kat.json pins happy-path
// round-trips + every DIDError code.

package com.elabify.core

data class DID(val network: String, val entityType: String, val identifier: String)

enum class DIDErrorCode {
    MALFORMED,
    EXTRA_COLONS,
    EMPTY_COMPONENT,
}

class DIDError(val code: DIDErrorCode, message: String) : RuntimeException(message)

fun parseDID(s: String): DID {
    val parts = s.split(":")
    if (parts.size > 5) {
        throw DIDError(DIDErrorCode.EXTRA_COLONS, "parseDID: too many colons in \"$s\"")
    }
    if (parts.size != 5) {
        throw DIDError(DIDErrorCode.MALFORMED, "parseDID: \"$s\" is not a did:elabify DID")
    }
    if (parts[0] != "did" || parts[1] != "elabify") {
        throw DIDError(DIDErrorCode.MALFORMED, "parseDID: \"$s\" must start with did:elabify:")
    }
    val network = parts[2]
    val entityType = parts[3]
    val identifier = parts[4]
    if (network.isEmpty() || entityType.isEmpty() || identifier.isEmpty()) {
        throw DIDError(DIDErrorCode.EMPTY_COMPONENT, "parseDID: \"$s\" has an empty component")
    }
    return DID(network, entityType, identifier)
}

fun formatDID(did: DID): String = "did:elabify:${did.network}:${did.entityType}:${did.identifier}"

File: bindings/swift/Sources/ElabifyCore/Did.swift
// did:elabify parsing and formatting. Byte-equivalent with @elabify/core's
// src/did.ts. Grammar (exactly five colon-separated components):
//
//   did:elabify:<network>:<entityType>:<identifier>
//
// Cross-binding equivalence: test-vectors/did.kat.json pins happy-path
// round-trips + every DIDError code.

import Foundation

public struct DID: Equatable, Hashable {
    public let network: String
    public let entityType: String
    public let identifier: String
    public init(network: String, entityType: String, identifier: String) {
        self.network = network
        self.entityType = entityType
        self.identifier = identifier
    }
}

public enum DIDError: Error, Equatable {
    case malformed
    case extraColons
    case emptyComponent
}

public func parseDID(_ s: String) throws -> DID {
    let parts = s.split(separator: ":", omittingEmptySubsequences: false).map(String.init)
    if parts.count > 5 { throw DIDError.extraColons }
    if parts.count != 5 { throw DIDError.malformed }
    if parts[0] != "did" || parts[1] != "elabify" { throw DIDError.malformed }
    let network = parts[2]
    let entityType = parts[3]
    let identifier = parts[4]
    if network.isEmpty || entityType.isEmpty || identifier.isEmpty {
        throw DIDError.emptyComponent
    }
    return DID(network: network, entityType: entityType, identifier: identifier)
}

public func formatDID(_ did: DID) -> String {
    return "did:elabify:" + did.network + ":" + did.entityType + ":" + did.identifier
}
